// @nuxt/cli/dist/cli-chunk6.js
/*!
 * @nuxt/cli v2.4.3 (c) 2016-2019

 * - All the amazing contributors
 * Released under the MIT License.
 * Website: https://nuxtjs.org
*/
'use strict';

require('path');
require('consola');
require('fs');
require('esm');
require('./cli-chunk2.js');
require('@nuxt/config');
require('boxen');
require('chalk');
require('pretty-bytes');
require('std-env');
const __chunk_3 = require('./cli-chunk3.js');

const build = {
  name: 'build',
  description: 'Compiles the application for production deployment',
  usage: 'build <dir>',
  options: {
    ...__chunk_3.common,
    analyze: {
      alias: 'a',
      type: 'boolean',
      description: 'Launch webpack-bundle-analyzer to optimize your bundles',
      prepare(cmd, options, argv) {
        // Analyze option
        options.build = options.build || {};
        if (argv.analyze && typeof options.build.analyze !== 'object') {
          options.build.analyze = true;
        }
      }
    },
    devtools: {
      type: 'boolean',
      default: false,
      description: 'Enable Vue devtools',
      prepare(cmd, options, argv) {
        options.vue = options.vue || {};
        options.vue.config = options.vue.config || {};
        if (argv.devtools) {
          options.vue.config.devtools = true;
        }
      }
    },
    generate: {
      type: 'boolean',
      default: true,
      description: 'Don\'t generate static version for SPA mode (useful for nuxt start)'
    },
    quiet: {
      alias: 'q',
      type: 'boolean',
      description: 'Disable output except for errors',
      prepare(cmd, options, argv) {
        // Silence output when using --quiet
        options.build = options.build || {};
        if (argv.quiet) {
          options.build.quiet = !!argv.quiet;
        }
      }
    },
    standalone: {
      type: 'boolean',
      default: false,
      description: 'Bundle all server dependencies (useful for nuxt-start)',
      prepare(cmd, options, argv) {
        if (argv.standalone) {
          options.build.standalone = true;
        }
      }
    }
  },
  async run(cmd) {
    const config$$1 = await cmd.getNuxtConfig({ dev: false });
    const nuxt = await cmd.getNuxt(config$$1);

    if (nuxt.options.mode !== 'spa' || cmd.argv.generate === false) {
      // Build only
      const builder = await cmd.getBuilder(nuxt);
      await builder.build();
    } else {
      // Build + Generate for static deployment
      const generator = await cmd.getGenerator(nuxt);
      await generator.generate({ build: true });
    }
  }
};

exports.default = build;
