// @nuxt/cli/dist/cli-chunk7.js
/*!
 * @nuxt/cli v2.4.3 (c) 2016-2019

 * - All the amazing contributors
 * Released under the MIT License.
 * Website: https://nuxtjs.org
*/
'use strict';

require('path');
require('consola');
require('fs');
require('esm');
const __chunk_2 = require('./cli-chunk2.js');
require('@nuxt/config');
require('boxen');
require('chalk');
require('pretty-bytes');
require('std-env');
const __chunk_3 = require('./cli-chunk3.js');

const generate = {
  name: 'generate',
  description: 'Generate a static web application (server-rendered)',
  usage: 'generate <dir>',
  options: {
    ...__chunk_3.common,
    build: {
      type: 'boolean',
      default: true,
      description: 'Only generate pages for dynamic routes. Nuxt has to be built once before using this option'
    },
    devtools: {
      type: 'boolean',
      default: false,
      description: 'Enable Vue devtools',
      prepare(cmd, options, argv) {
        options.vue = options.vue || {};
        options.vue.config = options.vue.config || {};
        if (argv.devtools) {
          options.vue.config.devtools = true;
        }
      }
    },
    modern: {
      ...__chunk_3.common.modern,
      description: 'Generate app in modern build (modern mode can be only client)',
      prepare(cmd, options, argv) {
        if (__chunk_2.normalizeArg(argv.modern)) {
          options.modern = 'client';
        }
      }
    }
  },
  async run(cmd) {
    const config$$1 = await cmd.getNuxtConfig({ dev: false });
    const nuxt = await cmd.getNuxt(config$$1);
    const generator = await cmd.getGenerator(nuxt);

    await generator.generate({
      init: true,
      build: cmd.argv.build
    });
  }
};

exports.default = generate;
